src/Symfony/Component/DependencyInjection/EnvProvider.php

```diff
@@ -0,0 +1,166 @@
+            $env = $_ENV[$name];
+        } elseif (false === ($env = getenv($name)) || null === $env) { // null is a possible value because of thread safety issues
+            if (!$this->container->hasParameter("env($name)")) {
+                throw new EnvNotFoundException($name);
+            }
+
+            if (null === $env = $this->container->getParameter("env($name)")) {
+                return;
+            }
+        }
+
+        if (!is_scalar($env)) {
+            throw new RuntimeException(sprintf('Non-scalar env var "%s" cannot be cast to %s.', $name, $prefix));
+        }
+
+        if ('string' === $prefix) {
+            return (string) $env;
+        }
+
+        if ('bool' === $prefix) {
+            return (bool) self::phpize($env);
+        }
+
+        if ('int' === $prefix) {
+            if (!is_numeric($env = self::phpize($env))) {
+                throw new RuntimeException(sprintf('Non-numeric env var "%s" cannot be cast to int.', $name));
+            }
+
+            return (int) $env;
+        }
+
+        if ('float' === $prefix) {
+            if (!is_numeric($env = self::phpize($env))) {
+                throw new RuntimeException(sprintf('Non-numeric env var "%s" cannot be cast to float.', $name));
+            }
+
+            return (float) $env;
+        }
+
+        if ('const' === $prefix) {
+            if (!defined($env)) {
+                throw new RuntimeException(sprintf('Env var "%s" maps to undefined constant "%s".', $name, $env));
+            }
+
+            return constant($name);
+        }
+
+        if ('base64' === $prefix) {
+            return base64_decode($env);
+        }
+
+        if ('json' === $prefix) {
+            $env = json_decode($env, true, JSON_BIGINT_AS_STRING);
+
+            if (JSON_ERROR_NONE !== json_last_error()) {
+                throw new RuntimeException(sprintf('Invalid JSON in env var "%s": '.json_last_error_msg(), $name));
+            }
+
+            if (!is_array($env)) {
+                throw new RuntimeException(sprintf('Invalid JSON env var "%s": array expected, %s given.', $name, gettype($env)));
+            }
+
+            return $env;
+        }
+
+        if ('resolve' === $prefix) {
+            return preg_replace_callback('/%%|%([^%\s]+)%/', function ($match) use ($name) {
+                if (!isset($match[1])) {
+                    return '%';
+                }
+                $value = $this->container->getParameter($match[1]);
+                if (!is_scalar($value)) {
+                    throw new RuntimeException(sprintf('Parameter "%s" found when resolving env var "%s" must be scalar, "%s" given.', $match[1], $name, gettype($value)));
+                }
+
+                return $value;
+            }, $env);
+        }
+
+        throw new RuntimeException(sprintf('Unsupported env var prefix "%s".', $prefix));
+    }
+
+    private static function phpize($value)
+    {
+        if (!class_exists(XmlUtils::class)) {
```

Comment:
Probably a smell that this class should be renamed, or at least some part of it extracted into a more generic class.

Reply:
not a blocker here, and no better idea :)